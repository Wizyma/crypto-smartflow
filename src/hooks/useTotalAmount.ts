import { useMemo } from 'react';

import useAmounts from 'hooks/useAmounts';
import { CryptoRsponse } from 'types/dataContract';

/**
 * @function useTotalAmount
 * @summary get total amount based on user wallet and crypto value
 *
 * @export useTotalAmount
 * @param {{
 *   cryptos: CryptoRsponse;
 *   wallet: { [key: string]: number };
 * }} {
 *   cryptos,
 *   wallet,
 * }
 * @returns {{ percentage: {[key: string]: number}, total: number }} Total amount
 */
export default function useTotalAmount({
  cryptos,
  wallet,
}: {
  cryptos: CryptoRsponse;
  wallet: { [key: string]: number };
}) {
  const amounts = useAmounts({ cryptos, wallet });
  const total = useMemo(() => Object.values(amounts).reduce((prev, next) => prev + next, 0), [amounts]);
  const percentage = useMemo(
    () =>
      Object.keys(amounts).map((key) => ({
        [key]: ((100 * amounts[key]) / total).toFixed(2),
      })),
    [amounts, total],
  );

  return {
    total: total.toFixed(2),
    percentage,
  };
}
