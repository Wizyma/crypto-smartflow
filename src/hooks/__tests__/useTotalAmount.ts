import { renderHook } from '@testing-library/react-hooks';

import useTotalAmount from 'hooks/useTotalAmount';

const dataTest = {
  cryptos: {
    BTC: { EUR: 8567 },
    ETH: { EUR: 200 },
    DASH: { EUR: 0.66 },
    BAT: { EUR: 2.57 },
    USDC: { EUR: 98 },
  },
  wallet: { BTC: 0.1005, ETH: 2.2349, DASH: 0.2527, BAT: 82.4851, USDC: 14.5099 },
};

test('it should return the total amount and percentage by crypto', () => {
  const { result } = renderHook(() => useTotalAmount(dataTest));
  expect(parseFloat(result.current.total)).toBe(2942.09);
  expect(result.current.percentage).toMatchSnapshot();
});
