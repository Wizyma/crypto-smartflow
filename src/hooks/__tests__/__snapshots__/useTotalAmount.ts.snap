// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should return the total amount and percentage by crypto 1`] = `
Array [
  Object {
    "BTC": "29.26",
  },
  Object {
    "ETH": "15.19",
  },
  Object {
    "DASH": "0.01",
  },
  Object {
    "BAT": "7.21",
  },
  Object {
    "USDC": "48.33",
  },
]
`;
